File: [0813][Medium][Largest_Sum_of_Averages]/Largest_Sum_of_Averages.java
// leetcode time     cost : 9 ms
// leetcode memory   cost : 37.8 MB 
// Time  Complexity: O(N*N*K)
// Space Complexity: O(N*K)
// solution 1, DFS
class Solution {
    public double largestSumOfAverages(int[] A, int K) {
        int n = A.length;
        /**
        dp[i][k]表示前i个数构成k个子数组时的最大平均值, 那么对于所有 0 <= j <= i-1
        dp[i][k] = Math.max(dp[i][k], dp[j][k-1] + (A[j+1] + ... + A[i+1]) / (i-j))
        **/
        double[][] dp = new double[n + 1][K + 1];
        double sum[] = new double[n + 1];
        for(int i = 1; i <= n; i++){
            sum[i] = sum[i - 1] + A[i - 1];
            dp[i][1] = sum[i] / i;
        }
        for(int i = 1; i <= n; i++){
            for(int k = 2; k <= K; k++){
                //由于要求每组非空，所以从k - 1开始
                for(int j = k - 1; j < i; j++){
                    dp[i][k] = Math.max(dp[i][k],dp[j][k - 1] + (sum[i] - sum[j]) / (i - j));
                }
            }
        }
        return dp[n][K];
    }
}

File: [0813][Medium][Largest_Sum_of_Averages]/Largest_Sum_of_Averages.py
# leetcode time     cost : 592 ms
# leetcode memory   cost : 13.6 MB 
# Time  Complexity: O(N*N*K)
# Space Complexity: O(N*K)
# solution 1.DP. 
# 设 dp(i, k) 表示数组 A 中从第 i 个元素到结尾 A[i:] 分成 k 个相邻的非空子数组，可以得到的最大分数
from typing import List
class Solution(object):
    def largestSumOfAverages(self, A, K):
        P = [0]
        for x in A: P.append(P[-1] + x)
        def average(i, j):
            return (P[j] - P[i]) / float(j - i)

        N = len(A)
        dp = [average(i, N) for i in range(N)]
        for k in range(K-1):
            for i in range(N):
                for j in range(i+1, N):
                    dp[i] = max(dp[i], average(i, j) + dp[j])

        return dp[0]

def main():
    A, K = [9,1,2,3,9],3      #expect is 20
    obj = Solution()
    res = obj.largestSumOfAverages(A, K)
    print("return value is ",res);
    
if __name__ =='__main__':
    main()   